src/modules/did/did.service.ts

```diff
@@ -358,10 +363,30 @@ export class DIDService implements OnModuleInit, OnModuleDestroy {
 
   private async syncDocuments() {
     this.logger.debug(`Beginning sync of DID Documents`);
-    const cachedDIDs = await this.didRepository.find({ select: ['id'] });
-    cachedDIDs.forEach(async (did) => {
+    const syncs = await this.latestDidSyncRepository.find({
+      order: { createdDate: 'DESC' },
+    });
+    const fromBlock = syncs.length > 0 ? syncs[0].block : 0;
+    const topBlock = await this.provider.getBlockNumber();
+    this.logger.debug(
+      `Getting DID update events from block ${fromBlock} to block ${topBlock}...`
+    );
+    const changedIdentities = await this.getChangedIdentities(
+      fromBlock,
+      topBlock
+    );
+    const didsToSynchronize = (
+      await this.didRepository.find({ select: ['id'] })
+    ).filter((doc) => {
+      const identity = doc.id.split(':')[3];
+      return changedIdentities.includes(identity);
+    });
+    didsToSynchronize.forEach(async (did) => {
+      this.logger.debug(`Synchronizing DID ${did.id}`);
       await this.pinDocument(did.id);
     });
+
+    await this.latestDidSyncRepository.save({ block: topBlock });
```

Comment:
Hey @JGiter, thanks for the PR. Really exciting that this code is being improved. 

I think this approach is flawed however. The issue that I see is that jobs are persisted in separate storage (Redis) and then processed asynchronously, but the `latestDidSync` `topBlock` is persisted synchronously. So we could have a situation like this:
1. A bunch of DID update jobs are queued and `lastestDidSync.block` is updated
2. The Redis storage fails or is wiped (it's in-memory, so definitely could happen) -> then the system won't update the DIDs in the lost jobs because `lastestDidSync.block` already ahead of these events.

I'm thinking an alternative approach to get around this problem would be to:
1. Mark updated DIDs as "stale":
  1. Query all of the "changedIdentities" since the last check (as you've done in this PR)
  2. Mark all of these DIDs as "invalid" (in the Postgres Entities). I think something [like this](https://stackoverflow.com/a/62125466) could be used to bulk update entities.
  3. Update `latestDidSync.block`
2. Query all of DID entities with invalid status and add a job to the queue (if it doesn't exist already) 
3. When done processing the job, update the DID to be "valid"

In this above, even if the job queue is reset/wiped, the DID entities will still have invalid status, so job for them can be re-added.

Also, we can change the `GET Did Doc` endpoint to synchronously query the RPC if the cached DID is invalid.

Reply:
Thanks. This is really good improvement